functions/src/storeTrackingData.ts

```diff
@@ -1,4 +1,110 @@
-exports.handler = async function(snap: any) {
+export {};
+const { google } = require('googleapis');
+const functions = require('firebase-functions');
+const GenderApi = require('gender-api.com-client');
+
+function getAppDate(appDate: any) {
+  const monthNames = [
+    'Jan',
+    'Feb',
+    'Mar',
+    'Apr',
+    'May',
+    'Jun',
+    'Jul',
+    'Aug',
+    'Sep',
+    'Oct',
+    'Nov',
+    'Dec',
+  ];
+  const newdate =
+    appDate.getDate() +
+    '-' +
+    monthNames[appDate.getMonth()] +
+    '-' +
+    appDate.getFullYear();
+  return newdate;
+}
+
+exports.handler = async function(snap: any, context: any, admin: any) {
   const email: string = snap.data()!.email;
-  console.log(email);
+  const name: string = snap.data()!.name;
+  const batch: string = snap.data()!.batch;
+  const userType: string = snap.data()!.userType;
+  const track: string = snap.data()!.track;
+  const source: string = snap.data()!.source;
+  const applicationTime: any = getAppDate(snap.createTime.toDate());
+  const docId = context.params.applicationId;
+  const batchId = context.params.batch;
+
+  const OAuth2 = google.auth.OAuth2;
+  const clientID = functions.config().spreadsheet.clientid;
+  const clientSecret = functions.config().spreadsheet.clientsecret;
+  const refreshToken = functions.config().spreadsheet.refreshtoken;
+  const genderApiKey = functions.config().spreadsheet.genderapikey;
+  const googleSheetID = functions.config().spreadsheet.id;
+
+  const oauth2Client = new OAuth2(
+    clientID, //client Id
+    clientSecret, // Client Secret
+    'https://developers.google.com/oauthplayground' // Redirect URL
+  );
+
+  oauth2Client.setCredentials({
+    refresh_token: refreshToken,
+  });
+
+  // api used: https://gender-api.com/en/clients
+  const genderApiClient = new GenderApi.Client(genderApiKey);
+  genderApiClient.getByFirstName(name, function(result: any) {
+    admin
+      .firestore()
+      .collection('batches')
+      .doc(batchId)
+      .collection('applications')
+      .doc(docId)
+      .update({
+        applicationDate: applicationTime,
+        gender: result.gender,
+      })
+      .then(function() {
+        console.log('Document successfully updated!');
+      })
+      .catch(function(error: any) {
+        console.log('Error adding document: ', error);
+      });
```

Comment:
We shouldn't mix concerns of a function here. I'd rather do this in the `handleFormSubmit` function. In this way, we have one spot in the code, where we insert data for an application.

Reply:
Ok I just realised we have this pull request still open..
So regarding this change, I tried making the change you requested and had also a discussion with lyu. Kind of forgot to have a talk with you regarding this..

So, first the `handleFormSubmit` function is triggered when the document with application data is submitted. Then once this document is created `storeTrackingData` function is triggered in which the genderAPI works on the submitted data (name field) to determine the gender. The gender gets determined and gets appended as new field to the document. 

Shifting this insertion code to `handleFormSubmit` is tricky because genderAPI needs the 'name' field to work on. So I don't think that here the strict separation of concerns would work. Even if there is a way to make it work, probably it would be too much work for this use case. 

Would like to know your views :)